crawl-ref/source/initfile.cc

```diff
@@ -5670,6 +5718,184 @@ bool parse_args(int argc, char **argv, bool rc_only)
     return true;
 }
 
+// Like MenuEntry, but only parse menu colours for the first _highlight_until
+// characters of each line (in this case, the option name but not its value).
+class EGP_MenuEntry : public MenuEntry
+{
+public:
+    EGP_MenuEntry(const string &txt = string(), MenuEntryLevel lev = MEL_ITEM,
+                  int qty = 0, int hotk = 0, int _highlight_until = INT_MAX)
+        : MenuEntry(txt, lev, qty, hotk), highlight_until(_highlight_until) { }
+
+    int highlight_colour() const override
+    {
+        return menu_colour(text.substr(0, highlight_until), "", tag);
+    }
+private:
+    int highlight_until;
+};
+
+class EGP_Menu : public Menu
+{
+public:
+    EGP_Menu(int _flags) : Menu(_flags)
+                           { set_title(new MenuEntry(base_title, MEL_TITLE));}
+
+    ~EGP_Menu()
+    {
+        deleteAll(all_items);
+        items.clear();
+    }
+
+    void add_entry_all(MenuEntry *entry)
+    {
+        add_entry(entry);
+        all_items.push_back(entry);
+    }
+
+    void filter_items()
+    {
+        string prompt = "Search for what? (regex, leave blank to show all)";
+        char select[1024];
+        auto old_search_pat = search_pat;
+        int new_hovered = -1;
+        vector<MenuEntry*> chosen_items;
+        while (1)
```

Comment:
```suggestion
        while (true)
```

Reply:
I changed it to `while (chosen_items.empty())` instead.